proposals/CodifiedVUs.adoc

```diff
@@ -0,0 +1,170 @@
+// Copyright 2023 The Khronos Group Inc.
+//
+// SPDX-License-Identifier: CC-BY-4.0
+
+= Codified VUs
+:toc: left
+:refpage: https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/
+:sectnums:
+
+In this proposal, a new procedural syntax for writing Valid Usage (VU)
+statements in the spec is introduced.
```

Comment:
Hm, I wonder how daring would be to do it the other way around, rather than inventing yet another language.

I.e. take the C++ layers and generate something reasonable into the spec. It would require some discipline though with the validation layer architecture and codestyle, which so far was little bit organic.

Reply:
Quite daring I'd say :)

So far, VVL doesn't seem very amenable to large-scale changes, and autogen for these is a smaller scale than that. It was suggested before though, that validation could live outside the spec (but probably still in this syntax).

It's quite hard to go from C++ to something simple to read I'd say. The other way around is much easier (even if it's still quite a big change)

Reply:
Either way feels like starting from scratch. Starting with C++ would though mean we end up with one language and one source instead of two.

C++ is simple to read to me. Well, depends on the writer. The examples in this proposal are quite stone-age C-istic , and not quite leveraging the more beautiful parts of Python anyway.

As I was saying in the forums, I think it would be great if the codified VUs were more semantic than procedural. And ideally more one-liney. E.g. I think most of the VUs should look something like `for_all( filter( pArray, has_desired_property ), elementIsValid );`, rather than a web of raw `for`s and `if`s.

Reply:
> BTW, is the part of this to still attempt to back-translate to human-readable language, or given the audience we can afford to stick to codified VUs only?

That's the assumption, yes, that we're all programmers so we can read simple code just fine.

> As I was saying in the forums, I think it would be great if the codified VUs were more semantic than procedural. And ideally more one-liney. E.g. I think most of the VUs should look something like all( filter( pArray, only_elements_with_this_property ), element_be_valid );, rather than a web of raw fors and ifs.

There's a balance to be had here and I think we can adjust over time. Requiring more of python (say lambdas) makes this technically more difficult (bear in mind that I'm tackling VVL code generation as well), but that aside it also means Vulkan developers would have to have some non-trivial familiarity with Python.

As much as I'd love to have that, I'm hesitant to add anything that makes VUs unreadable to people who don't know Python. You are right that the syntax is basic, and that was on purpose. Anyone can read `if x > 0:`, even if they've never seen Python before.

That said, I've already noticed some cases that become too verbose with this language, like the one before last in this file: https://github.com/KhronosGroup/Vulkan-Docs/pull/2064/commits/af37bc96975071abde37d3e43d2dda6558f7ebbd#diff-6c1b4ac11212d29616a39eb88d49ade5da7def6269874081730c261ae2f61562
which is exactly where I'd like to be able to use `any(...)`. I'm still considering how to best approach that, but definitely as a "next step", as this is already a big job.

Reply:
It is pretty much modern C++ too, so it is not that unreadable. Actually it may be more readable to people outside programming, because it looks more English or more like standard math expression: ∀elem∈Arr: elem = valid_elem

Reply:
I don't disagree with you, and again this'll be something I'll have to iterate on and add support for over time.

That said, this:

```
for elem in arr:
    require(valid(elem))
```

is literally the same as `∀elem∈Arr: valid(elem)` written in typable characters, so I don't believe it's _that_ bad.

Reply:
I don't much believe in iterating on stuff. People never return later to redo stuff unless it explodes, hehe.

It is not the same. `for` means "go over elements in arr one by one from first to last, and do something to it be it read or write access (and we are not telling which)". It is semantically something different from "Some predicate we supply later applies to every element of array". It just happens to be only procedurally same.

It is not so pronounced and visible in singular raw `for`, because the brain will make the translation in inperceivable half a second. But it is annoying if you open a page with 42 `for`s or interspersed with a web of (raw) `if`s.

It is not something I am pulling out of my ass, and it's not that a new thing. Maybe I should quote appropriate articles: https://www.fluentcpp.com/2017/01/05/the-importance-of-knowing-stl-algorithms/

> one of the main advantage algorithms bring is expressiveness, by raising the level of abstraction of code. That is to say they show **what** they do, rather than **how** they are implemented.
> [...]
> Don’t use  `for_each` for each problem
> [...]
> one specific case where `for_each` is effectively adapted: when **performing side effects**.
> [...]
> If you need to know whether all the elements of a collection satisfy a given predicate, use `std::all_of`.

https://youtu.be/W2tWOdzgXHA?t=131

> goal nr. 1: No raw loops
> [...]
> I would say, take all your coding guidelines, and replace them with the one goal. That's how important I think this one goal is!

Reply:
> Hm, I wonder how daring would be to do it the other way around, rather than inventing yet another language.

In some ways I think this is ideal, but I'm not sure how one would achieve it. I'd like to say this would be possible with clang tooling and perhaps special comments/annotations, but I don't have the knowledge nor experience with those tools to confidently say that's true.

One thing none of this helps with is testing. VVL is more stringent on requiring tests for newly added VUID checks, but this still relies on "human enforcement" at review time. While automating testing may be outside of the realm of possibility, I wonder if we could generate the spec VUID text from VVL test output (which in theory prints "every" VUID). This would
1. Enforce testing for _every_ VUID
2. Improve (hopefully) the quality of runtime VUID error messages (a longstanding goal of VVL), as they would be subject to the same scrutiny the spec is

Reply:
@ncesario-lunarg It helps testing in the way that it makes it nearly irrelevant. The tests largely test whether we manually translated the human-readable VU to code correctly. That largely falls if VU **is** code, right?

Reply:
> The tests largely test whether we manually translated the human-readable VU to code correctly.

Yes, but it also exercises the checks on a number of different drivers/platforms/etc. Even if the checks are 100% accurate and correct, I would say the checks are only correct in theory without actually running them on _some_ HW/drivers. That said, if it were possible, generating the spec from code does seem more appealing the more I think about it.

Reply:
On the subject of testing, it feels to me like if mock ICD could somewhat accurately represent different devices (i.e. respond to complex queries the same way the real driver does), you could rely on it for a lot of the testing (other than for GPU-AV of course). Like you could reduce your costs a lot by testing on different emulations of devices, and then, like, run a nightly test on real hardware.

It'll probably be a lot of work to make that happen though.

> That said, if it were possible, generating the spec from code does seem more appealing the more I think about it.

Glad to hear that :tada:

Did you see my presentation at the last F2F?

Reply:
> Glad to hear that 🎉

Just to clarify, this is w.r.t. to the _general idea_ of generating the spec from the C++ code that actually makes the runtime check. Which is somewhat ideal, but I'm not sure how one would achieve it.

I do think having VUs codified would yield some great benefits (like the ones you've highlighted at the F2F, which I did see ;)), but making that change this late in Vulkan 1.x's development still seems too disruptive IMHO.

Reply:
@ncesario-lunarg Not sure what VVL tests have to do with HW. VVL runs against API usage not against HW\driver. Conformace suite tests the HW\driver.